### PR TITLE
authorizer for dedicated deployment

### PR DESCRIPTION
**Pull request by PacificDou**
# Description

authorizer for dedicated deployment: check if api_key belongs to the dedicated workspace or not

List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

Tested locally

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

**Comment by PawelPeczek-Roboflow** [comment]
Looks ok, but I would like to know more details of the change

**Comment by PacificDou** [comment]
There is an import format check error.
Tried to reformat the file, but nothing happened...
```
(.venv) shuyang@shuyang-dev:~/inference/inference/core/interfaces/http$ black http_api.py 
All done! ✨ 🍰 ✨
1 file left unchanged.
```

File: inference/core/interfaces/http/http_api.py [hunk 1]
```diff
@@ -446,6 +451,70 @@ async def count_errors(request: Request, call_next):
...
+                    else dict()
+                )
+                api_key = req_params.get("api_key", None) or json_params.get(
+                    "api_key", None
+                )
+
+                if api_key not in cached_api_keys:
+                    try:
+                        workspace_url = (
+                            get_roboflow_workspace(api_key)
```
Comment (bigbitbus) [inline inference/core/interfaces/http/http_api.py:487]: If someone has the wrong API key (or say, a DDoS with a wrong API key), and they send many requests, this call to api.roboflow.com can result in rate-limiting, so maybe we should have a limit on how many such checks can be performed per minute by this server.

File: inference/core/interfaces/http/http_api.py [hunk 2]
```diff
@@ -446,6 +451,70 @@ async def count_errors(request: Request, call_next):
...
+                # check project_url
+                model_id = json_params.get("model_id", "")
+                project_url = (
+                    req_params.get("project", None)
+                    or json_params.get("project", None)
+                    or model_id.split("/")[0]
+                )
+                if project_url is not None and project_url not in cached_projects:
+                    try:
+                        _ = get_roboflow_dataset_type(
```
Comment (bigbitbus) [inline inference/core/interfaces/http/http_api.py:508]: Same comment as above, we should try to limit how many times a pod can query the api.roboflow.com server.

**Review by bigbitbus — COMMENTED** [review]
@PacificDou I left some comments.

**Comment by PawelPeczek-Roboflow** [comment]
@PacificDou that was `isort` complaining not black. I guess we should enable pre-commit hooks to solve this kind of issues

**Review by PawelPeczek-Roboflow — APPROVED** [review]

**Review by grzegorz-roboflow — APPROVED** [review]

File: inference/core/interfaces/http/http_api.py (hunk 1 above)
Comment (yeldarby) [reply inference/core/interfaces/http/http_api.py:487]: Agree responsibility for enforcement of the rate limit is on the backend but the problem would be that an external malicious actor could get this client's IP blocked by coercing it into getting rate limited by the backend.

**Pull request merged**